### PR TITLE
Tests for SageMaker update and archiving functionality

### PR DESCRIPTION
**Pull request by dbczumar**
This PR introduces additional tests to ensure that S3/SageMaker resources are not deleted prior to the completion of SageMaker endpoint updates.

### CODE REVIEW

File: mlflow/sagemaker/__init__.py [hunk 1]
```diff
@@ -380,8 +380,10 @@ def deploy(app_name, model_path, execution_role_arn=None, bucket=None, run_id=No
                          operation_status.message)
         else:
             raise MlflowException(
-                "The deployment operation failed with the following error message:"
-                " \"{error_message}\"".format(error_message=operation_status.message))
+                message=(
+                    "The deployment operation failed with the following error message:"
+                    " \"{error_message}\"".format(error_message=operation_status.message)),
+                error_code=PUBLIC)
```
Comment (mparkhe) [inline mlflow/sagemaker/__init__.py]: is `PUBLIC` a type of error code? Looks like it is API visibility type

**Review by mparkhe — COMMENTED** [review]
Please could you add some documentation in the mock classes as what is the expected behavior, status changes ... etc

File: mlflow/sagemaker/__init__.py (hunk 1 above)
Comment (dbczumar) [reply mlflow/sagemaker/__init__.py]: Good call - removed `PUBLIC` and reverted back to `INTERNAL_ERROR`, since everything else seemed misleading.

**Comment by dbczumar** [comment]
@mparkhe I changed the abstractions around endpoint status slightly; I think they're clearer now. Each endpoint has its most recent operation as an attribute. That operation has a time-dependent status. `EndpointOperations` are set by the `SageMakerBackend` rather than the `Endpoint` objects themselves, which is better than having `Endpoints` manage their own state.

**Comment by mparkhe** [comment]
LGTM.

Not sure what is happening with Travis. Please remember to resolve those issues before merging.

**Pull request merged**